[ENH] Shift centers of mass outside of clusters to closest in-cluster voxel

clane9 opened the pull request:
Closes #3278.

Changes proposed in this pull request:

- For binary clusters whose center of mass falls outside of the cluster, report the nearest within-cluster voxel instead.
- Don't filter locally constant "maxima" in the binary cluster case.
- Vectorize some operations.

bthirion reviewed (commented):
The change LGTM, thx. Can you add a whatsnew entry ?

NicolasGensollen reviewed (commented):
Thanks for working on this @clane9 !
The changes LGTM. I just have a couple small suggestions below.
As pointed out by @bthirion, could you add a line in `doc/changes/latest.rst`. Since this is your first contribution to Nilearn, you'll also need to add your name and profile link to `doc/changes/names.rst`.

clane9 commented on nilearn/reporting/_get_clusters_table.py:
```diff
@@ -73,40 +73,76 @@ def _identify_subpeaks(data):
     return ijk, vals
 
 
+def _cluster_nearest_neighbor(ijk, labels_index, labeled):
```
Good suggestion! Indeed there was a typo.

clane9 commented on nilearn/reporting/_get_clusters_table.py:
```diff
@@ -73,40 +73,76 @@ def _identify_subpeaks(data):
+    """
+    labels = labeled[labeled > 0]
+    cluster_ijk = np.array(labeled.nonzero()).T
+    nbrs = []
```
Good point!

clane9 commented:
Thanks for the quick review @bthirion, @NicolasGensollen! I think I've addressed your comments and fixed a typo in the nearest neighbor function. Lmk if you need anything else.

NicolasGensollen reviewed (approved):
Great! LGTM! 👍 
Thanks for addressing my points @clane9 !

tsalo commented on nilearn/reporting/_get_clusters_table.py:
```diff
@@ -73,40 +73,76 @@ def _identify_subpeaks(data):
+    if np.any(subpeaks_outside_cluster):
         warnings.warn(
             "Attention: At least one of the (sub)peaks falls outside of the "
             "cluster body."
```
```suggestion
            "cluster body. Identifying the nearest in-cluster voxel."
```

tsalo reviewed (changes requested):
This really looks great. Thank you for opening this! I just have a few small requests/notes.

clane9 commented:
Thanks for the feedback @tsalo! I think I've incorporated your suggestions.

tsalo reviewed (approved):
Thank you! LGTM.

clane9 commented on nilearn/reporting/_get_clusters_table.py:
```diff
@@ -73,40 +73,76 @@ def _identify_subpeaks(data):
+    Returns
+    -------
+    nbrs : `numpy.ndarray`
+        (n_pts, 3) nearest neighbor points.
```
Ah, right. Ok I added the tags for the rest of the parameters. I decided not to go through and update all the other docstrings, since that seemed out of scope for this PR. But Lmk if you think it's a good opportunity to do that.

tsalo replied:
Thank you for doing that. As a general rule, we try to limit formatting updates to functions that are directly affected by a given PR, so we don't clutter up the git blame, so I think that was the right choice.

bthirion reviewed (approved):
LGTM, thx.

NicolasGensollen commented:
Great! Merging since we have three approvals...
